### ISSUE TITLE
How to flush SDK data for testing purpose?

### ISSUE THREAD

**Issue by matas-ramanauskas-revel** [DataDog/dd-sdk-ios#1618]
### The issue
I'm writing tests for logging Library that uses this sdk. I want to write a few logs and flush them then test with API if they are in the correct place and accessable. The only way to do it I found so far is flushAndDeinitialize which is hidden behind "DD_SDK_COMPILED_FOR_TESTING". I set this flag in my app, but it seems not to reach inside the library still. Is there a way to make it work for testing purposes?

### Reproduction
```swift
// package.swift
    targets: [
        .executableTarget(
            name: "test-dd",
            dependencies: [
                .product(name: "DatadogCore", package: "dd-sdk-ios"),
                .product(name: "DatadogLogs", package: "dd-sdk-ios")
            ],
            path: "Sources",
            swiftSettings: [.define("DD_SDK_COMPILED_FOR_TESTING")]
        ),
```

```swift
// main.swift
...
Datadog.flushAndDeinitialize()
```

---

#### Datadog SDK version:

.upToNextMajor(from: "2.0.0") 

#### Last working Datadog SDK version:

none

#### Dependency Manager:

SPM

#### Xcode version:

15.1

#### Swift version:

5.8

#### Deployment Target:

iOS ipad (10 gen)

#### macOS version:

Ventura

**Comment by ncreated** [comment]
Hello @matas-ramanauskas-revel 👋. It isn't possible to set [`swiftSettings`](https://developer.apple.com/documentation/packagedescription/target/swiftsettings) for target's dependency in SPM - like you noticed, it is available on [`Target`](https://developer.apple.com/documentation/packagedescription/package/targets), not [`Target.Dependency`](https://developer.apple.com/documentation/packagedescription/target/dependency). I think this design makes sense, as one may want to use the same package as dependency to multiple targets and if it is configured differently each time, SPM won't be able to resolve it.

We use `DD_SDK_COMPILED_FOR_TESTING` compilation condition on purpose: this API is not meant to be publicly available and we only use it internally in our tests. If you seek on basing your tests on Datadog iOS SDK's internals, you can leverage `@testable import` and call [the internal implementation of flush method](https://github.com/DataDog/dd-sdk-ios/blob/a43c63dd422ffd050ed24a1ac1dfd7b156f063c2/DatadogCore/Sources/Datadog.swift#L491):
```swift
@testable import DatadogCore

Datadog.internalFlushAndDeinitialize()
```

While this will work in your tests, we strongly discourage using it in production app. Keep in mind that this API is internal, so we may decide to change its implementation or rename it without any notice.

There is no public API available that would let you flush the SDK's state. Closer to that is `Datadog.clearAllData()` but that only deletes all unsent files from the device.

I hope this answers the question and helps you 🙂. Closing this issue as answered, feel free to re-open with more context or further question on flushing.

**Comment by matas-ramanauskas-revel** [comment]
Thank you for fast response. We will not use this in prod for sure, just for testing. 
And as I see, it still needs to wait a few secs after doing the flush, since even ending test instantly after flushing does still fail to send logs. So like 2-5s thread sleep is still in order sadly.

**Comment by matas-ramanauskas-revel** [comment]
Hmm.. The flush does not seem to work correctly or am I missing something? 
This still does not work:
```swift
logger.info("Hello")
Datadog.internalFlushAndDeinitialize()
```

Only thing that is working is:
```swift
logger.info("Hello")
try? await Task.sleep(nanoseconds: 30 * 1_000_000_000)
Datadog.internalFlushAndDeinitialize()
```

What would be correct practice to test if the data is being sent? Is the seep mandatory? Are there any other ways to make library send data to datadog on demand?

**Comment by ncreated** [comment]
@matas-ramanauskas-revel I did some extra test and indeed, the `.internalFlushAndDeinitialize()` will not work without `DD_SDK_COMPILED_FOR_TESTING` set. I consider this redundant, hence I opened https://github.com/DataDog/dd-sdk-ios/pull/1628 to fix this.

> What would be correct practice to test if the data is being sent? Is the seep mandatory? Are there any other ways to make library send data to datadog on demand?

The `sleep()` would be mandatory if you want to use only official (public) APIs. You can configure [`uploadFrequency: .frequent`](https://docs.datadoghq.com/real_user_monitoring/mobile_and_tv_monitoring/advanced_configuration/ios/?tab=swift#initialization-parameters) to force SDK to send data faster. Also, setting [`batchProcessingLevel: .high`](https://github.com/DataDog/dd-sdk-ios/blob/36862590661791e1ac6a0d61fbaf863c94cea97d/DatadogCore/Sources/Datadog.swift#L126-L129) will cause more intensive (and quicker) uploads.

Except the workaround through internal `.internalFlushAndDeinitialize()` that I described earlier there is no other way for forcing data uploads from the SDK. There is no public API for this.

@matas-ramanauskas-revel I wonder what is the exact use case you're implementing? Why exactly do you want to assert data uploads from Datadog SDK as part of your tests?

**Comment by matas-ramanauskas-revel** [comment]
Thank you! Will wait for a new version with the fix for implementing these tests then ^^ 
@ncreated As for exact use case: we have a library, that is responsible for logging it has a couple of logging mechanisms, one of them is uploading uploading logs to datadog. We have multiple different possibilities for logging, with lots of custom attributes and with some of them having additional rules setup in datadog dashboard itself. Currently everything is working as expected, however we haven't done any changes to this for over a year. Some new requriments appeared for us and also some canged happened with datadog itself (current version used is 1.11.1. We'd like to upgrade sdk to latest and then start using other additional features too. However we need to make sure that all the current functionallity is not broken. That means we need our changes and changes that happen between ios-skd version do not change anything. And if it happens to know what changed and be able to put it back into place. Hence we wrote tests for all the custom logging and want to check if they appear in datadog in the same exact form and place.

**Comment by ncreated** [comment]
@matas-ramanauskas-revel thanks for extra context, it makes sense. What you can even think of is fetching logs back into your tests with our [`GET /logs/events`](https://docs.datadoghq.com/api/latest/logs/#get-a-list-of-logs), although that would imply even more `sleep()` to anticipate latency in DD logs processing.

The https://github.com/DataDog/dd-sdk-ios/pull/1628 was just merged, so you can try it out right away if installing SDK from `develop` branch (for testing purpose). Otherwise, you can wait for next release (no ETA yet).

Let me know if I can help more on this topic. Otherwise consider closing this issue or we will close it with next release.

**Comment by ncreated** [comment]
@matas-ramanauskas-revel [**2.7.0**](https://github.com/DataDog/dd-sdk-ios/releases/tag/2.7.0) is out and it includes the fix #1628. Closing this issue 🙂
